### ISSUE TITLE
kubelet fails to unmount CSI volumes, floods log with errors and has to be restarted

### ISSUE THREAD

**Issue by pavilalopes** [kubernetes/kubernetes#101378]
#### What happened:

On pod termination, kubelet sometimes fails to unmount CSI volumes with the following error:
```
E0422 13:28:12.214727    1861 reconciler.go:193] operationExecutor.UnmountVolume failed (controllerAttachDetachEnabled true) for volume "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "dc3f11b7-d671-464c-a734-65e1760ed7df" (UID: "dc3f11b7-d671-464c-a734-65e1760ed7df") : UnmountVolume.NewUnmounter failed for volume "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "dc3f11b7-d671-464c-a734-65e1760ed7df" (UID: "dc3f11b7-d671-464c-a734-65e1760ed7df") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/dc3f11b7-d671-464c-a734-65e1760ed7df/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/dc3f11b7-d671-464c-a734-65e1760ed7df/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/vol_data.json]: open /var/lib/kubelet/pods/dc3f11b7-d671-464c-a734-65e1760ed7df/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/vol_data.json: no such file or directory
```

Once this bug is triggered the Pod is stuck in Terminating state and the above error message starts repeating hundreds of times per minute in the log.
The os-level mounts themselves are not removed, as reported by `mount` on the host:
```
10.115.38.19:6789,10.115.38.14:6789,10.115.38.17:6789:/volumes/csi/csi-vol-97172dfa-9e15-11eb-8ecd-26ee97ae76aa/44564d2b-e10b-4f25-a668-701552549ca0 on /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/globalmount type ceph (rw,relatime,name=csi-cephfs-node,secret=<hidden>,acl,mds_namespace=storefs,_netdev)
10.115.38.19:6789,10.115.38.14:6789,10.115.38.17:6789:/volumes/csi/csi-vol-97172dfa-9e15-11eb-8ecd-26ee97ae76aa/44564d2b-e10b-4f25-a668-701552549ca0 on /var/lib/kubelet/pods/dc3f11b7-d671-464c-a734-65e1760ed7df/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/mount type ceph (rw,relatime,name=csi-cephfs-node,secret=<hidden>,acl,mds_namespace=storefs,_netdev)
```

Only solution after this point seems to be to restart the kubelet and clean up the pod by hand (by force-deleting the Pods and unmounting the volume on the host).

PVCs are being supplied by Ceph Rook, but see "Anything else we need to know?" section.

#### What you expected to happen:

kubelet should be able to unmount CSI volumes on pod termination.

#### How to reproduce it (as minimally and precisely as possible):

We launch a couple of hundred Pods and then stop them. About 5-10% of them fail to terminate with this problem. Pods can have anywhere from 1 to 13 containers, with every container mounting the same PVC ("cephfs-config" in the example below). Pod declarations are all some variation of this:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: tracker-group
  namespace: default
spec:
  selector:
    matchLabels:
      app: tracker-group
  template:
    metadata:
      labels:
        app: tracker-group
    spec:
      containers:
      - command: [...] #redacted
        env: [...] #redacted
        image: "my-image" #redacted
        imagePullPolicy: Always
        name: tracker-crop-producer
        volumeMounts:
        - mountPath: /app/params
          name: tracker-crop-producer--params
          readOnly: true
        - mountPath: /app/config
          name: persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm
          readOnly: true
      - command: [...] #redacted
        env: [...] #redacted
        image: "my-image" #redacted
        imagePullPolicy: Always
        name: tracker-work-assigner
        volumeMounts:
        - mountPath: /app/params
          name: tracker-work-assigner--params
          readOnly: true
        - mountPath: /app/config
          name: persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm
          readOnly: true
      dnsPolicy: ClusterFirstWithHostNet
      hostNetwork: true
      volumes:
      - configMap:
          name: tracker-crop-producer--params
        name: tracker-crop-producer--params
      - configMap:
          name: tracker-work-assigner--params
        name: tracker-work-assigner--params
      - name: persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm
        persistentVolumeClaim:
          claimName: cephfs-config
          readOnly: true
```

#### Anything else we need to know?:

This may be the same bug as #85280 which was closed by the bot without resolution. (That bug does not mention how the problem might occur naturally, only how to trigger it by intentionally messing with the /var/lib/kubelet/pods/... directory.)

While we are experiencing this problem with Ceph Rook, we found [this bug](https://github.com/longhorn/longhorn/issues/2368) filed in the longhorn project, showing the exact same problem happening with their CSI.

#### Environment:
- Kubernetes version (use `kubectl version`):
   ```
   Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"6b1d87acf3c8253c123756b9e61dac642678305f", GitTreeState:"clean", BuildDate:"2021-03-18T01:10:43Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
   Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"6b1d87acf3c8253c123756b9e61dac642678305f", GitTreeState:"clean", BuildDate:"2021-03-18T01:02:01Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
   ```
- Cloud provider or hardware configuration: Bare metal 10 node cluster
- OS (e.g: `cat /etc/os-release`): Ubuntu 20.04.2 LTS
- Kernel (e.g. `uname -a`): Linux appl-1-8 5.4.0-72-generic #80-Ubuntu SMP Mon Apr 12 17:35:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: kubeadm
- Network plugin and version (if this is a network-related bug):
- Others: Rook Ceph operator v1.5.9 (Ceph v15.2.9)

**Comment by niulechuan** [comment]
looks like the vol_data.json file was missing in /var/lib/kubelet/pods/..., did it deleted by mistaken?

**Comment by pavilalopes** [comment]
Yes, I know that is the problem from the log message. We checked and the file was indeed missing.
If you are asking if the file was deleted by some external process, then no.

**Comment by niulechuan** [comment]
@pavilalopes Could you find any error msg like `failed to delete volume path` in kubelet logs?

**Comment by pavilalopes** [comment]
Searching verbatim for "failed to delete" does not find anything in the logs.

I'm going to try to collect a set of log messages for the case where the unmount succeeded, and one where it failed. I'll post it here shortly.

**Comment by niulechuan** [comment]
Okay, thank you!

**Comment by pavilalopes** [comment]
Ok, here are the logs from 2 pods form the same run, executing on the same node.

First a success case:
```
I0422 11:29:36.752077    1425 operation_generator.go:556] MountVolume.WaitForAttach entering for volume "pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "tracking-detector-group-general-3-9-10-11-6dbfcd6876-ztv66" (UID: "4df251bb-12ce-4993-a258-91dcfab178ac") DevicePath "csi-d9838258de66f38a9ff756cb63dcdfd90e1e2be60012eb4344cb0f1cacbb5abd"
I0422 11:29:37.304727    1425 operation_generator.go:565] MountVolume.WaitForAttach succeeded for volume "pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "tracking-detector-group-general-3-9-10-11-6dbfcd6876-ztv66" (UID: "4df251bb-12ce-4993-a258-91dcfab178ac") DevicePath "csi-d9838258de66f38a9ff756cb63dcdfd90e1e2be60012eb4344cb0f1cacbb5abd"
I0422 11:33:25.861173    1425 reconciler.go:196] operationExecutor.UnmountVolume started for volume "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "4df251bb-12ce-4993-a258-91dcfab178ac" (UID: "4df251bb-12ce-4993-a258-91dcfab178ac")
I0422 11:33:25.865880    1425 operation_generator.go:797] UnmountVolume.TearDown succeeded for volume "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa" (OuterVolumeSpecName: "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm") pod "4df251bb-12ce-4993-a258-91dcfab178ac" (UID: "4df251bb-12ce-4993-a258-91dcfab178ac"). InnerVolumeSpecName "pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df". PluginName "kubernetes.io/csi", VolumeGidValue ""
```

Then a failure case:
```
I0422 11:28:37.396184    1425 operation_generator.go:556] MountVolume.WaitForAttach entering for volume "pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "products-in-hand-detection-cameras-53-55-56-5c6d8b579b-rq9z2" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d") DevicePath "csi-d9838258de66f38a9ff756cb63dcdfd90e1e2be60012eb4344cb0f1cacbb5abd"
I0422 11:28:37.988921    1425 operation_generator.go:565] MountVolume.WaitForAttach succeeded for volume "pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "products-in-hand-detection-cameras-53-55-56-5c6d8b579b-rq9z2" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d") DevicePath "csi-d9838258de66f38a9ff756cb63dcdfd90e1e2be60012eb4344cb0f1cacbb5abd"
I0422 11:32:50.900321    1425 reconciler.go:196] operationExecutor.UnmountVolume started for volume "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "5e11dfee-0b9a-46ed-9933-6b26530a405d" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d")
I0422 11:32:50.904555    1425 operation_generator.go:797] UnmountVolume.TearDown succeeded for volume "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa" (OuterVolumeSpecName: "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm") pod "5e11dfee-0b9a-46ed-9933-6b26530a405d" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d"). InnerVolumeSpecName "pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df". PluginName "kubernetes.io/csi", VolumeGidValue ""
I0422 11:33:16.473621    1425 operation_generator.go:556] MountVolume.WaitForAttach entering for volume "pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "products-in-hand-detection-cameras-53-55-56-5c6d8b579b-rq9z2" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d") DevicePath "csi-d9838258de66f38a9ff756cb63dcdfd90e1e2be60012eb4344cb0f1cacbb5abd"
I0422 11:33:17.001114    1425 operation_generator.go:565] MountVolume.WaitForAttach succeeded for volume "pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "products-in-hand-detection-cameras-53-55-56-5c6d8b579b-rq9z2" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d") DevicePath "csi-d9838258de66f38a9ff756cb63dcdfd90e1e2be60012eb4344cb0f1cacbb5abd"
E0422 11:33:25.868742    1425 reconciler.go:193] operationExecutor.UnmountVolume failed (controllerAttachDetachEnabled true) for volume "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "5e11dfee-0b9a-46ed-9933-6b26530a405d" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d") : UnmountVolume.NewUnmounter failed for volume "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "5e11dfee-0b9a-46ed-9933-6b26530a405d" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/5e11dfee-0b9a-46ed-9933-6b26530a405d/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/5e11dfee-0b9a-46ed-9933-6b26530a405d/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/vol_data.json]: open /var/lib/kubelet/pods/5e11dfee-0b9a-46ed-9933-6b26530a405d/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/vol_data.json: no such file or directory
E0422 11:33:25.972132    1425 reconciler.go:193] operationExecutor.UnmountVolume failed (controllerAttachDetachEnabled true) for volume "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "5e11dfee-0b9a-46ed-9933-6b26530a405d" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d") : UnmountVolume.NewUnmounter failed for volume "persistentvolumemount-6wg4ywt2xwwgkyygj6tyo2xhrm" (UniqueName: "kubernetes.io/csi/rook-ceph.cephfs.csi.ceph.com^0001-0009-rook-ceph-0000000000000001-97172dfa-9e15-11eb-8ecd-26ee97ae76aa") pod "5e11dfee-0b9a-46ed-9933-6b26530a405d" (UID: "5e11dfee-0b9a-46ed-9933-6b26530a405d") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/5e11dfee-0b9a-46ed-9933-6b26530a405d/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/5e11dfee-0b9a-46ed-9933-6b26530a405d/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/vol_data.json]: open /var/lib/kubelet/pods/5e11dfee-0b9a-46ed-9933-6b26530a405d/volumes/kubernetes.io~csi/pvc-b27ae1f5-bfa6-4951-aebb-ff33280796df/vol_data.json: no such file or directory
... message repeats until kubelet is killed ...
```

The extra "MountVolume.WaitForAttach" operations after the volume has already been unmounted seem strange to me.

I grepped for references to the CSI UniqueName. Obviously there are more messages for these pods in the log, let me know if you would like a more complete log (though they get quite big....)

**Comment by niulechuan** [comment]
I will dig into, for now the log is enough.

**Comment by niulechuan** [comment]
Sorry to disturb you on the weekend, I have to need the complete log of kubelet, or 11:25:00-11:35:00, yesterday I checked the code, I need more details to locate the root cause. You can get it to me on weekdays.

**Comment by pavilalopes** [comment]
Here is the unabridged log for those 10 min:
[20210422_1125.log.gz](https://github.com/kubernetes/kubernetes/files/6375502/20210422_1125.log.gz)

**Comment by niulechuan** [comment]
👌

**Comment by aviorma** [comment]
[csi-cephfsplugin.log](https://github.com/kubernetes/kubernetes/files/6397402/csi-cephfsplugin.log)


[kubelet.log](https://github.com/kubernetes/kubernetes/files/6397404/kubelet.log)
Having the same issues with ROOK + CEPHFS
adding logs for investigation

**Comment by timmy59100** [comment]
Same with longhorn, presumably after deleting volume data on note.

**Comment by eliekoren** [comment]
+1

**Comment by SerialVelocity** [comment]
Same here. It might be related to the upgrade from rook 1.5.x to 1.6.2, but not sure. Could it be something they added to deal with K8s 1.21 but it causes issues for 1.20?

Kubernetes should handle this gracefully no matter what though

**Comment by msau42** [comment]
I think this is the same issue as https://github.com/kubernetes/kubernetes/issues/101911

/assign @msau42 @gnufied 
/triage accepted
@kubernetes/sig-storage-bugs

**Comment by i1337m** [comment]
Any progress regarding this bug?

**Comment by eliekoren** [comment]
Same question

**Comment by gohilankit** [comment]
> The fix for orphaned pod volume cleanup is in 1.22 and being backported to 1.21, 1.20 and 1.19. #102576

@msau - Do you know which patch versions of 1.21 & 1.20 has the fix? I couldn't find the PR #102576 in changelog for 1.20 & 1.21.

**Comment by vadasambar** [comment]
> > The fix for orphaned pod volume cleanup is in 1.22 and being backported to 1.21, 1.20 and 1.19. #102576
> 
> @msau - Do you know which patch versions of 1.21 & 1.20 has the fix? I couldn't find the PR #102576 in changelog for 1.20 & 1.21.

I can see the code from the PR [in release-1.20 branch](https://github.com/kubernetes/kubernetes/blob/0d6dc14051612afe6a9b45708edef60f30a8cd4a/pkg/kubelet/kubelet_volumes.go#L119)

**Comment by thamurepo** [comment]
**Am facing same issue , is there any fix, Am using v-1.24 , any ETA**

Apr 16 04:27:19 ip-10-69-80-134.ec2.internal kubelet[5471]: E0416 04:27:19.105849    5471 reconciler.go:208] "operationExecutor.UnmountVolume failed (controllerAttachDetachEnabled true) for volume \"shared-logging-storage\" (UniqueName: \"kubernetes.io/csi/nfs.csi.k8s.io^10.69.80.161/shared-logging-storage\") pod \"f8c88ab3-71be-4cd5-911a-c9647242fe60\" (UID: \"f8c88ab3-71be-4cd5-911a-c9647242fe60\") : UnmountVolume.NewUnmounter failed for volume \"shared-logging-storage\" (UniqueName: \"kubernetes.io/csi/nfs.csi.k8s.io^10.69.80.161/shared-logging-storage\") pod \"f8c88ab3-71be-4cd5-911a-c9647242fe60\" (UID: \"f8c88ab3-71be-4cd5-911a-c9647242fe60\") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/f8c88ab3-71be-4cd5-911a-c9647242fe60/volumes/kubernetes.io~csi/shared-logging-storage-nfs-pv/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/f8c88ab3-71be-4cd5-911a-c9647242fe60/volumes/kubernetes.io~csi/shared-logging-storage-nfs-pv/vol_data.json]: open /var/lib/kubelet/pods/f8c88ab3-71be-4cd5-911a-c9647242fe60/volumes/kubernetes.io~csi/shared-logging-storage-nfs-pv/vol_data.json: no such file or directory" err="UnmountVolume.NewUnmounter failed for volume \"shared-logging-storage\" (UniqueName: \"kubernetes.io/csi/nfs.csi.k8s.io^10.69.80.161/shared-logging-storage\") pod \"f8c88ab3-71be-4cd5-911a-c9647242fe60\" (UID: \"f8c88ab3-71be-4cd5-911a-c9647242fe60\") : kubernetes.io/csi: unmounter failed to load volume data file [/var/lib/kubelet/pods/f8c88ab3-71be-4cd5-911a-c9647242fe60/volumes/kubernetes.io~csi/shared-logging-storage-nfs-pv/mount]: kubernetes.io/csi: failed to open volume data file [/var/lib/kubelet/pods/f8c88ab3-71be-4cd5-911a-c9647242fe60/volumes/kubernetes.io~csi/shared-logging-storage-nfs-pv/vol_data.json]: open /var/lib/kubelet/pods/f8c88ab3-71be-4cd5-911a-c9647242fe60/volumes/kubernetes.io~csi/shared-logging-storage-nfs-pv/vol_data.json: no such file or directory"
